core/imds/src/test/java/software/amazon/awssdk/imds/Ec2MetadataWithApacheClientTest.java

```diff
@@ -0,0 +1,389 @@
+        stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE)).willReturn(aResponse().withBody("{}").withStatus(401)));
+
+        assertThatThrownBy(() -> {
+            Ec2Metadata ec2Metadata =
+                Ec2Metadata.builder().httpClient(httpClient).endpoint(URI.create("http://localhost:8080")).build();
+            MetadataResponse metadataResponse = ec2Metadata.get("/latest/meta-data/ami-id");
+        }).hasMessageContaining("Exceeded maximum number of retries.")
+          .isInstanceOf(SdkClientException.class);
+    }
+
+    @Test
+    public void getToken_failedThriceWith403() throws IOException {
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).willReturn(aResponse().withStatus(403)));
+        stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE)).willReturn(aResponse().withBody("{}")));
+
+        assertThatThrownBy(() -> {
+            Ec2Metadata ec2Metadata = Ec2Metadata.builder().httpClient(httpClient).endpoint(URI.create("http://localhost:8080")).build();
+            MetadataResponse metadataResponse = ec2Metadata.get("/latest/meta-data/ami-id");
+        }).hasMessageContaining("token")
+          .isInstanceOf(SdkClientException.class);
+    }
+
+    @Test
+    public void getToken_failedThriceWith401() throws IOException {
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).willReturn(aResponse().withStatus(401)));
+        stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE)).willReturn(aResponse().withBody("{}")));
+
+        assertThatThrownBy(() -> {
+            Ec2Metadata ec2Metadata = Ec2Metadata.builder().httpClient(httpClient).endpoint(URI.create("http://localhost:8080")).build();
+            MetadataResponse metadataResponse = ec2Metadata.get("/latest/meta-data/ami-id");
+        }).hasMessageContaining("Exceeded maximum number of retries.")
+          .isInstanceOf(SdkClientException.class);
+        WireMock.verify(putRequestedFor(urlPathEqualTo(TOKEN_RESOURCE_PATH)).withHeader(EC2_METADATA_TOKEN_TTL_HEADER, equalTo("21600")));
+
+    }
+
+    @Test
+    public void getToken_failedThriceWithFixedDelay() throws IOException {
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).willReturn(aResponse().withFixedDelay(Integer.MAX_VALUE)));
+        stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE)).willReturn(aResponse().withBody("{}")));
+
+        assertThatThrownBy(() -> {
+            Ec2Metadata ec2Metadata = Ec2Metadata.builder().httpClient(httpClient).endpoint(URI.create("http://localhost:8080")).build();
+            MetadataResponse metadataResponse = ec2Metadata.get("/latest/meta-data/ami-id");
+        }).hasMessageContaining("Exceeded maximum number of retries.")
+          .isInstanceOf(SdkClientException.class);
+
+        WireMock.verify(putRequestedFor(urlPathEqualTo(TOKEN_RESOURCE_PATH)).withHeader(EC2_METADATA_TOKEN_TTL_HEADER, equalTo("21600")));
+    }
+
+    @Test
+    public void getToken_failedOnceWith401_shouldSucceedOnSecondAttempt_Apache() throws IOException {
```

Comment:
How are these test cases specific to Apache client ? Or in other word why do we have individual failure test for Apache client.
Is retry handling done specific to Apache client

Reply:
The test cases have the Apache Client passed it in the builder. The test cases are copied over so to ensure that the Apache Client too has passed all the scenarios practically.